server/src/main/java/org/elasticsearch/transport/RemoteClusterService.java

```diff
@@ -304,12 +307,58 @@ private synchronized void updateSkipUnavailable(String clusterAlias, Boolean ski
         }
     }
 
-    public void updateRemoteClusterCredentials(Settings settings) {
-        remoteClusterCredentialsManager.updateClusterCredentials(settings);
+    public synchronized void updateRemoteClusterCredentials(Supplier<Settings> settingsSupplier, ActionListener<Void> listener) {
+        updateRemoteClusterCredentials(settingsSupplier.get(), listener);
     }
 
-    public RemoteClusterCredentialsManager getRemoteClusterCredentialsManager() {
-        return remoteClusterCredentialsManager;
+    public synchronized void updateRemoteClusterCredentials(Settings settings, ActionListener<Void> listener) {
+        final UpdateRemoteClusterCredentialsResult result = remoteClusterCredentialsManager.updateClusterCredentials(settings);
+        // We only need to rebuild connections when a credential was newly added or removed for a cluster alias, not if the credential
+        // value was updated. Therefore, only consider added or removed aliases
+        final int totalConnectionsToRebuild = result.addedClusterAliases().size() + result.removedClusterAliases().size();
+        if (totalConnectionsToRebuild == 0) {
+            logger.debug("no connection rebuilding required after credentials update");
+            listener.onResponse(null);
+            return;
+        }
+        logger.info("rebuilding [{}] connections after credentials update", totalConnectionsToRebuild);
+        final GroupedActionListener<Void> groupedListener = new GroupedActionListener<>(
+            totalConnectionsToRebuild,
+            listener.map(responses -> {
+                logger.info("rebuild complete for [{}] connections after credentials update", responses.size());
+                return null;
+            })
+        );
+        for (var clusterAlias : result.addedClusterAliases()) {
+            maybeRebuildConnectionOnCredentialsChange(clusterAlias, settings, groupedListener);
+        }
+        for (var clusterAlias : result.removedClusterAliases()) {
+            maybeRebuildConnectionOnCredentialsChange(clusterAlias, settings, groupedListener);
+        }
+    }
+
+    private void maybeRebuildConnectionOnCredentialsChange(
+        String clusterAlias,
+        Settings settings,
+        GroupedActionListener<Void> groupedListener
+    ) {
+        if (remoteClusters.containsKey(clusterAlias)) {
+            updateRemoteCluster(clusterAlias, settings, true, ActionListener.wrap(status -> {
```

Comment:
does this handle tearing down a connection after removing a credential ?

Reply:
If a connection exists, and we remove a credential, `updateRemoteCluster` handles re-building it. If the connection does not exist, that means cluster settings for it were removed, and `updateRemoteCluster` will tear it down on the regular settings update call that removed the cluster settings. So if a connection does not exist anymore, and but a credential does still and is removed, there is nothing left to tear down. Let me know if I misinterpreted the question!